File: Classes/NKJMovieComposer.m
//
//  NKJMovieComposer.m
//
//  Created by nakajijapan.
//  Copyright 2014 nakajijapan. All rights reserved.
//

#import "NKJMovieComposer.h"

@interface NKJMovieComposer()
@end

@implementation NKJMovieComposer

#pragma mark - Initialize
- (id)init
{
    self = [super init];
    if (self) {

        // AVMutableVideoCompositionLayerInstruction's List
        _layerInstructions = [NSMutableArray array];

        // AVMutableComposition
        _mixComposition = [AVMutableComposition composition];
        AVMutableCompositionTrack *compositionVideoTrack;
        compositionVideoTrack = [_mixComposition addMutableTrackWithMediaType:AVMediaTypeVideo
                                                             preferredTrackID:kCMPersistentTrackID_Invalid];

        // AVMutableVideoComposition
        _videoComposition = [AVMutableVideoComposition videoComposition];
        [_videoComposition setRenderSize:CGSizeMake(640, 640)];
        [_videoComposition setFrameDuration:CMTimeMake(1, 24)];

    }

    return self;
}

#pragma mark - Cover Video
- (void)coverVideoWithURL:(NSURL*)movieURL scale:(CGAffineTransform)scale transform:(CGAffineTransform)transform
{
    AVURLAsset *videoAsset = [[AVURLAsset alloc] initWithURL:movieURL options:nil];
    AVMutableCompositionTrack *compositionVideoTrack;
    AVAssetTrack *videoTrack = [[videoAsset tracksWithMediaType:AVMediaTypeVideo] objectAtIndex:0];

    compositionVideoTrack = [_mixComposition addMutableTrackWithMediaType:AVMediaTypeVideo
                                                         preferredTrackID:kCMPersistentTrackID_Invalid];
    [compositionVideoTrack insertTimeRange:CMTimeRangeMake(kCMTimeZero, videoAsset.duration)
                                   ofTrack:videoTrack
                                    atTime:kCMTimeZero
                                     error:nil];
    [compositionVideoTrack setPreferredTransform:[videoTrack preferredTransform]];
    
    AVMutableVideoCompositionLayerInstruction *layerInstruction;
    layerInstruction = [AVMutableVideoCompositionLayerInstruction
                         videoCompositionLayerInstructionWithAssetTrack:compositionVideoTrack];
    
    [layerInstruction setTransform:CGAffineTransformConcat(scale, transform) atTime:kCMTimeZero];
    
    // Hide
    [layerInstruction setOpacity:0 atTime:_currentTimeDuration];

    [_layerInstructions addObject:layerInstruction];
}

#pragma mark - Add Video
- (AVMutableVideoCompositionLayerInstruction *)addVideoWithURL:(NSURL *)movieURL
{
    AVURLAsset *videoAsset = [[AVURLAsset alloc] initWithURL:movieURL options:nil];
    AVMutableCompositionTrack *compositionVideoTrack;
    AVAssetTrack *videoTrack = [[videoAsset tracksWithMediaType:AVMediaTypeVideo] objectAtIndex:0];
    
    compositionVideoTrack = [_mixComposition addMutableTrackWithMediaType:AVMediaTypeVideo
                                                         preferredTrackID:kCMPersistentTrackID_Invalid];
    [compositionVideoTrack insertTimeRange:CMTimeRangeMake(kCMTimeZero, videoAsset.duration)
                                   ofTrack:videoTrack
                                    atTime:_currentTimeDuration
                                     error:nil];
    [compositionVideoTrack setPreferredTransform:[[[videoAsset tracksWithMediaType:AVMediaTypeVideo] objectAtIndex:0] preferredTransform]];

    AVMutableCompositionTrack *compositionAudioTrack = [_mixComposition addMutableTrackWithMediaType:AVMediaTypeAudio
                                                                                   preferredTrackID:kCMPersistentTrackID_Invalid];
    [compositionAudioTrack insertTimeRange:CMTimeRangeMake(kCMTimeZero, videoAsset.duration)
                                   ofTrack:[[videoAsset tracksWithMediaType:AVMediaTypeAudio] objectAtIndex:0]
                                    atTime:_currentTimeDuration error:nil];
    
    _currentTimeDuration = [_mixComposition duration];

    // Add Layer Instruction
    AVMutableVideoCompositionLayerInstruction *layerInstruction;
    layerInstruction = [AVMutableVideoCompositionLayerInstruction
                        videoCompositionLayerInstructionWithAssetTrack:compositionVideoTrack];
    // Hide
    [layerInstruction setOpacity:0 atTime:_currentTimeDuration];

    [_layerInstructions addObject:layerInstruction];

    return layerInstruction;
}

#pragma mark - Export
- (AVAssetExportSession*)readyToComposeVideoWithFilePath: (NSString*)composedMoviePath
{
    if ([composedMoviePath isEqualToString:@""]) {
        return nil;
    }
    
    // create instruction
    _instruction = [AVMutableVideoCompositionInstruction videoCompositionInstruction];
    [_instruction setTimeRange:CMTimeRangeMake(kCMTimeZero, [_mixComposition duration])];

    [_videoComposition setInstructions: [NSArray arrayWithObject: _instruction]];
    [_instruction setLayerInstructions: [[_layerInstructions reverseObjectEnumerator] allObjects]];
    
    // generate AVAssetExportSession based on the composition
    _assetExportSession = [[AVAssetExportSession alloc] initWithAsset:_mixComposition presetName:AVAssetExportPreset1280x720];
    [_assetExportSession setVideoComposition: _videoComposition];
    [_assetExportSession setOutputFileType:AVFileTypeQuickTimeMovie];
    [_assetExportSession setOutputURL:[NSURL fileURLWithPath:composedMoviePath]];
    [_assetExportSession setShouldOptimizeForNetworkUse:YES];
    
    // delete file
    if ([[NSFileManager defaultManager] fileExistsAtPath:composedMoviePath]) {
        [[NSFileManager defaultManager] removeItemAtPath:composedMoviePath error:nil];
    }

    return _assetExportSession;
}

@end

File: Classes/NKJMovieComposer.swift
//
//  NKJMovieComposer.m
//
//  Created by nakajijapan.
//  Copyright 2014 nakajijapan. All rights reserved.
//

import Foundation
import AVFoundation
import CoreMedia

public class NKJMovieComposer {
    
    public var mixComposition: AVMutableComposition
    public var instruction: AVMutableVideoCompositionInstruction!
    public var videoComposition: AVMutableVideoComposition
    public var assetExportSession: AVAssetExportSession!
    public var currentTimeDuration: CMTime = kCMTimeZero
    public var layerInstructions: NSMutableArray
    
    public init() {
        
        // AVMutableVideoCompositionLayerInstruction's List
        self.layerInstructions = NSMutableArray()
        
        // AVMutableComposition
        self.mixComposition = AVMutableComposition()
        
        // AVMutableVideoComposition
        self.videoComposition = AVMutableVideoComposition()
        self.videoComposition.renderSize = CGSize(width: 640, height: 640)
        self.videoComposition.frameDuration = CMTimeMake(1, 24)
        
    }
    
    
    // Add Video
    public func addVideo(movieURL: NSURL!) -> AVMutableVideoCompositionLayerInstruction! {
        
        let videoAsset = AVURLAsset(URL:movieURL, options:nil)
        var compositionVideoTrack: AVMutableCompositionTrack!
        var compositionAudioTrack: AVMutableCompositionTrack!
        let videoTrack = videoAsset.tracksWithMediaType(AVMediaTypeVideo)[0] 
        
        compositionVideoTrack = self.mixComposition.addMutableTrackWithMediaType(AVMediaTypeVideo, preferredTrackID: 0)
        do {
            try compositionVideoTrack.insertTimeRange(
                CMTimeRange(start: kCMTimeZero, duration: videoAsset.duration),
                ofTrack: videoTrack,
                atTime: self.currentTimeDuration)
        } catch _ {
        }
        compositionVideoTrack.preferredTransform = videoTrack.preferredTransform
        
        compositionAudioTrack = self.mixComposition.addMutableTrackWithMediaType(AVMediaTypeAudio, preferredTrackID: 0)
        do {
            try compositionAudioTrack.insertTimeRange(
                CMTimeRange(start: kCMTimeZero, duration: videoAsset.duration),
                ofTrack: videoAsset.tracksWithMediaType(AVMediaTypeAudio)[0] ,
                atTime: self.currentTimeDuration)
        } catch _ {
        }
        
        self.currentTimeDuration = self.mixComposition.duration
        
        // Add Layer Instruction
        let layerInstruction = AVMutableVideoCompositionLayerInstruction(assetTrack: compositionVideoTrack)
        
        // Hide
        layerInstruction.setOpacity(0.0, atTime: self.currentTimeDuration)
        self.layerInstructions.addObject(layerInstruction)
        
        return layerInstruction
    }
    
    // Cover Video
    public func coverVideo(movieURL: NSURL!, scale: CGAffineTransform, transform: CGAffineTransform) -> AVMutableVideoCompositionLayerInstruction {
        
        let videoAsset = AVURLAsset(URL:movieURL, options:nil)
        var compositionVideoTrack:AVMutableCompositionTrack!
        let videoTrack = videoAsset.tracksWithMediaType(AVMediaTypeVideo)[0] 
        
        compositionVideoTrack = self.mixComposition.addMutableTrackWithMediaType(AVMediaTypeVideo, preferredTrackID: CMPersistentTrackID(kCMPersistentTrackID_Invalid))
        do {
            try compositionVideoTrack.insertTimeRange(
                CMTimeRange(start: kCMTimeZero, duration: videoAsset.duration),
                ofTrack: videoTrack,
                atTime: kCMTimeZero)
        } catch _ {
        }
        compositionVideoTrack.preferredTransform = videoTrack.preferredTransform
        
        
        var layerInstruction:AVMutableVideoCompositionLayerInstruction
        layerInstruction = AVMutableVideoCompositionLayerInstruction(assetTrack: compositionVideoTrack)
        layerInstruction.setTransform(CGAffineTransformConcat(scale, transform), atTime: kCMTimeZero)
        
        // Hide
        layerInstruction.setOpacity(0.0, atTime: self.currentTimeDuration)
        self.layerInstructions.addObject(layerInstruction)
        
        return layerInstruction
    }
    
    // Export
    public func readyToComposeVideo(composedMoviePath: String!) -> AVAssetExportSession! {
        
        if composedMoviePath == nil {
            return nil
        }
        
        // create instruction
        self.instruction = AVMutableVideoCompositionInstruction()
        self.instruction.timeRange = CMTimeRange(start: kCMTimeZero, duration: self.mixComposition.duration)
        
        self.videoComposition.instructions = [self.instruction]
        self.instruction.layerInstructions = self.layerInstructions.reverseObjectEnumerator().allObjects
        
        // generate AVAssetExportSession based on the composition
        self.assetExportSession = AVAssetExportSession(asset: self.mixComposition, presetName: AVAssetExportPreset1280x720)
        self.assetExportSession.videoComposition = self.videoComposition
        self.assetExportSession.outputFileType = AVFileTypeQuickTimeMovie
        self.assetExportSession.outputURL = NSURL(fileURLWithPath: composedMoviePath)
        self.assetExportSession.shouldOptimizeForNetworkUse = true
        
        // delete file
        if NSFileManager.defaultManager().fileExistsAtPath(composedMoviePath) {
            do {
                try NSFileManager.defaultManager().removeItemAtPath(composedMoviePath)
            } catch _ {
            }
        }
        return self.assetExportSession
    }
}
